// @dharmaprotocol/dharma.js/dist/lib/src/loan/agreement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var wrappers_1 = require("../wrappers");
var Agreement = /** @class */ (function () {
    function Agreement(dharma, params, data) {
        this.dharma = dharma;
        this.params = params;
        this.data = data;
    }
    Agreement.prototype.getAgreementId = function () {
        return new wrappers_1.DebtOrderDataWrapper(this.data).getIssuanceCommitmentHash();
    };
    Agreement.prototype.toJSON = function () {
        return {
            kernelVersion: this.data.kernelVersion,
            issuanceVersion: this.data.issuanceVersion,
            principalAmount: this.data.principalAmount.toString(),
            principalToken: this.data.principalToken,
            debtor: this.data.debtor,
            debtorFee: this.data.debtorFee.toString(),
            creditor: this.data.creditor,
            creditorFee: this.data.creditorFee.toString(),
            relayer: this.data.relayer,
            relayerFee: this.data.relayerFee.toString(),
            underwriter: this.data.underwriter,
            underwriterFee: this.data.underwriterFee.toString(),
            underwriterRiskRating: this.data.underwriterRiskRating.toString(),
            termsContract: this.data.termsContract,
            termsContractParameters: this.data.termsContractParameters,
            expirationTimestampInSec: this.data.expirationTimestampInSec.toString(),
            salt: this.data.salt.toString(),
            debtorSignature: this.data.debtorSignature,
            creditorSignature: this.data.creditorSignature,
            underwriterSignature: this.data.underwriterSignature,
        };
    };
    return Agreement;
}());
exports.Agreement = Agreement;
